src/snarks/groth16/mpc_utils.tcc

```diff
@@ -0,0 +1,212 @@
+            qap.C_in_Lagrange_basis[i];
+
+        A_i_g1[i] = tau_eval_g1.evaluate_from_lagrange_factors(A_i_in_lagrange);
+        B_i_g1[i] = tau_eval_g1.evaluate_from_lagrange_factors(B_i_in_lagrange);
+        B_i_g2[i] = tau_eval_g2.evaluate_from_lagrange_factors(B_i_in_lagrange);
+
+        G1 beta_A_at_t =
+            beta_tau_eval.evaluate_from_lagrange_factors(A_i_in_lagrange);
+        G1 alpha_B_at_t =
+            alpha_tau_eval.evaluate_from_lagrange_factors(B_i_in_lagrange);
+        G1 C_at_t = tau_eval_g1.evaluate_from_lagrange_factors(C_i_in_lagrange);
+
+        ABC_i_g1[i] = beta_A_at_t + alpha_B_at_t + C_at_t;
+    }
+
+    assert(num_variables + 1 == A_i_g1.size());
+    assert(num_variables + 1 == B_i_g1.size());
+    assert(num_variables + 1 == B_i_g2.size());
+    assert(num_variables + 1 == ABC_i_g1.size());
+
+    // TODO: Consider dropping those entries we know will not be used
+    // by this circuit and using sparse vectors where it makes sense
+    // (as is done for B_i's in r1cs_gg_ppzksnark_proving_key).
+
+    return srs_mpc_layer_L1<ppT>(
+        std::move(t_x_pow_i),
+        std::move(A_i_g1),
+        std::move(B_i_g1),
+        std::move(B_i_g2),
+        std::move(ABC_i_g1));
+}
+
+template<typename ppT>
+libsnark::r1cs_gg_ppzksnark_keypair<ppT> mpc_dummy_layer2(
+    srs_powersoftau &&pot,
+    srs_mpc_layer_L1<ppT> &&layer1,
+    const libff::Fr<ppT> &delta,
+    libsnark::r1cs_constraint_system<libff::Fr<ppT>> &&cs,
+    const libsnark::qap_instance<libff::Fr<ppT>> &qap)
+{
+    using Fr = libff::Fr<ppT>;
+    using G1 = libff::G1<ppT>;
+    using G2 = libff::G2<ppT>;
+
+    const Fr delta_inverse = delta.inverse();
+
+    // { H_i } = { [ t(x) . x^i / delta ]_1 } i = 0 .. n-2
+    libff::G1_vector<ppT> T_tau_powers_over_delta_g1(
+        layer1.T_tau_powers_g1.size());
+    for (size_t i = 0; i < layer1.T_tau_powers_g1.size(); ++i) {
+        T_tau_powers_over_delta_g1[i] =
+            delta_inverse * layer1.T_tau_powers_g1[i];
+    }
+
+    // ABC in verification key includes 1 + num_inputs terms.
+    // ABC/delta in prover key includes the remaining (num_variables -
+    // num_inputs) terms.
+    const size_t num_orig_ABC = layer1.ABC_g1.size();
+    const size_t num_variables = qap.num_variables();
+    const size_t num_inputs = qap.num_inputs();
+    const size_t num_L_elements = num_variables - num_inputs;
+
+    assert(num_orig_ABC == num_variables + 1);
+
+    // { ([B_i]_2, [B_i]_1) } i = 0 .. num_orig_ABC
+    std::vector<libsnark::knowledge_commitment<G2, G1>> B_i(num_orig_ABC);
+    for (size_t i = 0; i < num_orig_ABC; ++i) {
+        B_i[i] = libsnark::knowledge_commitment<G2, G1>(
+            layer1.B_g2[i], layer1.B_g1[i]);
+    }
+
+    assert(B_i.size() == qap.num_variables() + 1);
+
+    // { L_i } = [ { ABC_i / delta } ]_1, i = l+1 .. num_variables
+    libff::G1_vector<ppT> L_g1(num_L_elements);
+    for (size_t i = num_inputs + 1 ; i < num_variables + 1; ++i) {
```

Comment:
I think the build is complaining because of the extra space before `;` in this for:
`for (size_t i = num_inputs + 1 ; i < num_variables + 1; ++i) {` -> for (size_t i = num_inputs + 1; i < num_variables + 1; ++i) {

Reply:
Fixed in next commit

Reply:
Thanks for fixing that.